### DISCUSSION TITLE
AMS reader with an L&G E450 provided by "STW | Energie Klagenfurt (Austria)"

### DISCUSSION THREAD

**Discussion by fkroepfl** [UtilitechAS/amsreader-firmware#802] [Show and tell]
Below you'll see the screenshot after the initial Go/NoGo test of the AMS reader with an L&G E45 from "STW | Energie Klagenfurt" in Austria.

**Contact:**
Energie Klagenfurt
Smart Meter
St. Veiter Straße 31
A-9020 Klagenfurt am Wörthersee

T: +43 463 521-2900
E: smartmeter@stw.at
W: https://www.energieklagenfurt.at/stromnetz/smart-meter/
L: https://maps.app.goo.gl/1cypAy6qnARtjBPs9

**M-Bus interface:**
- Wired M-Bus interface
- Can be read out with an Mbus slave (eg.: https://de.aliexpress.com/item/1005004542854566.html)
- Standard EN 13757-2: 2005, DLMS/COSEM protocol
- Transmission rate 2,400 bps
- Output interval: 5 sec
- Encryption (in accordance with dlms/COSEM Security Suite 0, Policy 3) of the data output via the unidirectional customer interface. 

<img width="1164" alt="Bildschirmfoto 2024-06-01 um 11 57 36" src="https://github.com/UtilitechAS/amsreader-firmware/assets/6333880/c63a2f02-01c8-4547-9110-0d120481752c">
<img width="1164" alt="Bildschirmfoto 2024-06-01 um 11 57 45" src="https://github.com/UtilitechAS/amsreader-firmware/assets/6333880/de7ec35a-800d-48da-bdd6-34ea63adad0c">

[Kundenschnittstelle L&G  2.0.pdf](https://github.com/user-attachments/files/15522997/Kundenschnittstelle.L.G.2.0.pdf)

**Comment by fkroepfl** [comment]
**Additional info:**
Note: Allegedly the raw data packets are not complete according to the Mbus specification. Regardless of this, the payload should start from the position with '0204'.

`decrypted block: 355 bytes`

**Raw data:**
```
020f010f
020412002809060008190900ff0f02120000
020412002809060008190900ff0f01120000
020412000109060000600100ff0f02120000
020412000809060000010000ff0f02120000
020412000309060100010700ff0f02120000
020412000309060100020700ff0f02120000
020412000309060100010800ff0f02120000
020412000309060100020800ff0f02120000
020412000309060100030700ff0f02120000
020412000309060100040700ff0f02120000
020412000309060100030800ff0f02120000
020412000309060100040800ff0f02120000
0204120003090601001f0700ff0f02120000
020412000309060100330700ff0f02120000
020412000309060100470700ff0f02120000
09060008190900ff
09083632383534323235
090c07e8051c020e0523ff800080
060000074a
0600000000
0600bb1951
06001bf3db
0600000202
0600000000
06004cea5a
06000fb2ab
1200ae
1200c2
1201f5
1201f5
```

**Payload translated to XML:**
```
<Structure Qty="0F" >
  <Array Qty="0F" >
    <Structure Qty="04" >
      <UInt16 Value="0028" />
      <!-- 0.8.25.9.0.255 -->
      <OctetString Value="0008190900FF" />
      <Int8 Value="02" />
      <UInt16 Value="0000" />
    </Structure>
    <Structure Qty="04" >
      <UInt16 Value="0028" />
      <!-- 0.8.25.9.0.255 -->
      <OctetString Value="0008190900FF" />
      <Int8 Value="01" />
      <UInt16 Value="0000" />
    </Structure>
    <Structure Qty="04" >
      <UInt16 Value="0001" />
      <!-- 0.0.96.1.0.255 -->
      <OctetString Value="0000600100FF" />
      <Int8 Value="02" />
      <UInt16 Value="0000" />
    </Structure>
    <Structure Qty="04" >
      <UInt16 Value="0008" />
      <!-- 0.0.1.0.0.255 -->
      <OctetString Value="0000010000FF" />
      <Int8 Value="02" />
      <UInt16 Value="0000" />
    </Structure>
    <Structure Qty="04" >
      <UInt16 Value="0003" />
      <!-- 1.0.1.7.0.255 -->
      <OctetString Value="0100010700FF" />
      <Int8 Value="02" />
      <UInt16 Value="0000" />
    </Structure>
    <Structure Qty="04" >
      <UInt16 Value="0003" />
      <!-- 1.0.2.7.0.255 -->
      <OctetString Value="0100020700FF" />
      <Int8 Value="02" />
      <UInt16 Value="0000" />
    </Structure>
    <Structure Qty="04" >
      <UInt16 Value="0003" />
      <!-- 1.0.1.8.0.255 -->
      <OctetString Value="0100010800FF" />
      <Int8 Value="02" />
      <UInt16 Value="0000" />
    </Structure>
    <Structure Qty="04" >
      <UInt16 Value="0003" />
      <!-- 1.0.2.8.0.255 -->
      <OctetString Value="0100020800FF" />
      <Int8 Value="02" />
      <UInt16 Value="0000" />
    </Structure>
    <Structure Qty="04" >
      <UInt16 Value="0003" />
      <!-- 1.0.3.7.0.255 -->
      <OctetString Value="0100030700FF" />
      <Int8 Value="02" />
      <UInt16 Value="0000" />
    </Structure>
    <Structure Qty="04" >
      <UInt16 Value="0003" />
      <!-- 1.0.4.7.0.255 -->
      <OctetString Value="0100040700FF" />
      <Int8 Value="02" />
      <UInt16 Value="0000" />
    </Structure>
    <Structure Qty="04" >
      <UInt16 Value="0003" />
      <!-- 1.0.3.8.0.255 -->
      <OctetString Value="0100030800FF" />
      <Int8 Value="02" />
      <UInt16 Value="0000" />
    </Structure>
    <Structure Qty="04" >
      <UInt16 Value="0003" />
      <!-- 1.0.4.8.0.255 -->
      <OctetString Value="0100040800FF" />
      <Int8 Value="02" />
      <UInt16 Value="0000" />
    </Structure>
    <Structure Qty="04" >
      <UInt16 Value="0003" />
      <!-- 1.0.31.7.0.255 -->
      <OctetString Value="01001F0700FF" />
      <Int8 Value="02" />
      <UInt16 Value="0000" />
    </Structure>
    <Structure Qty="04" >
      <UInt16 Value="0003" />
      <!-- 1.0.51.7.0.255 -->
      <OctetString Value="0100330700FF" />
      <Int8 Value="02" />
      <UInt16 Value="0000" />
    </Structure>
    <Structure Qty="04" >
      <UInt16 Value="0003" />
      <!-- 1.0.71.7.0.255 -->
      <OctetString Value="0100470700FF" />
      <Int8 Value="02" />
      <UInt16 Value="0000" />
    </Structure>
  </Array>
  <!-- 0.8.25.9.0.255 -->
  <OctetString Value="0008190900FF" />
  <!-- 62854225 -->
  <OctetString Value="3632383534323235" />
  <!-- 28.05.2024 14:05:35 -->
  <OctetString Value="07E8051C020E0523FF800080" />
  <UInt32 Value="0000074A" />
  <UInt32 Value="00000000" />
  <UInt32 Value="00BB1951" />
  <UInt32 Value="001BF3DB" />
  <UInt32 Value="00000202" />
  <UInt32 Value="00000000" />
  <UInt32 Value="004CEA5A" />
  <UInt32 Value="000FB2AB" />
  <UInt16 Value="00AE" />
  <UInt16 Value="00C2" />
  <UInt16 Value="01F5" />
</Structure>
<!--
Extra data:
12 01 F5
-->
```

**Comment by fkroepfl** [comment]
**Question:** Can all OBIS data transmitted by the smart meter be displayed in the UI?

**Reply by ArnieO** [reply]
Which of them are not displayed?

**Reply by fkroepfl** [reply]
As I am still a newbie, it is of course possible that I have overlooked something, but at the moment it looks like this to me:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/cipo/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/cipo/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<meta charset=utf-8>
<meta charset=utf-8>
<meta charset=utf-8>
<meta charset=utf-8>
<meta charset=utf-8>
<meta charset=utf-8>
<meta charset=utf-8>
<meta charset=utf-8>
<meta charset=utf-8>
<meta charset=utf-8>
<meta charset=utf-8>
<meta charset=utf-8>
<meta charset=utf-8>
<meta charset=utf-8>

<body link="#0563C1" vlink="#954F72">


OBIS   Code | description | unit | alias | UI presence
-- | -- | -- | -- | --
0-0:96.1.0 | Meter serial number |   |   | info
0-0:1.0.0 | Meter date   & time |   |   | ?
1-0:1.7.0 | Positive active instantaneous power | kW | P+ | home
1-0:2.7.0 | Negative active instantaneous power | kW | P- | home
1-0:1.8.0 | Positive active energy, total | kWh | A+ | home
1-0:2.8.0 | Negative active energy, total | kWh | A- | home
1-0:3.7.0 | Positive reactive instantaneous power | kW | P+ | ?
1-0:4.7.0 | Negative reactive instantaneous power | kW | P- | ?
1-0:3.8.0 | Positive reactive energy, total | kvarh | Q+ | ?
1-0:4.8.0 | Negative reactive energy, total | kvarh | Q- | ?
1-0:31.7.0 | Instantaneous current in phase I | A | I (L1) | current   gauge?
1-0:51.7.0 | Instantaneous current in phase II | A | I (L2) | current gauge?
1-0:71.7.0 | Instantaneous current in phase III | A | I (L3) | current gauge?



</body>

</html>

It may not have been connected long enough to display the current consumption of the individual phase, or?

**Comment by ArnieO** [comment]
OK - a lot of new information here, thank you! And sorry for not having seen the attached document yesterday. 
If I understand correctly, it is I(L1), I(L2) and I(L3) that you are missing (they are not shown in your GUI screenshots). 

To help understand the reason for that, could you run a [Tenet debug](https://www.amsleser.no/module/ets_blog/blog?id_post=24), and post the result?

**Reply by fkroepfl** [reply]
I'll be happy to try it the next time I'm back on site. The meter is installed in my brother-in-law's property. 🤗

**Reply by fkroepfl** [reply]
> If I understand correctly, it is I(L1), I(L2) and I(L3) that you are missing (they are not shown in your GUI screenshots).

Yes, I think that would/should be the case, shouldn't it?

**Reply by ArnieO** [reply]
Good, looking forward to Telnet Debug printout (please post as text, not screenshot). 

> Yes, I think that would/should be the case, shouldn't it?

Absolutely, they should be there, and shown in GUI. And I don't see any reason why they are not there. They seem to be sent with the correct OBIS codes: 31.7.0/51.7.0/71.7.0. So we need to find out what the issue is here. 

We can se from your screenshots that meter serial number is read correctly, and so is Power and Energy. 

Just to have the complete picture: What about reactive power and energy (which are not very interesting for most of us, but should be part of the payload), can you see them? You might need to activate a tile in the config.

**Reply by fkroepfl** [reply]
> Just to have the complete picture: What about reactive power and energy (which are not very interesting for most of us, but should be part of the payload), can you see them? You might need to activate a tile in the config.

If I have understood the topic of reactive power/reactive energy correctly, the following applies:
- If a site has a high reactive power, i.e. if a lot of electricity is wasted, more electricity must flow to provide the same power.
- The result is therefore an unnecessarily high energy bill due to energy losses.

In this respect, there are two reasons why the values for reactive power/reactive energy should be displayed and - even better - not only displayed but also evaluated.

A) For the transparency of the received data telegrams, i.e. I can see everything the smart meter delivers in the GUI. This could be included as control information on the info page.

B) The evaluation of the relationship between active power and reactive power or active energy and reactive energy backed up with expertise. This could be used to show in the dashboard whether the measured household is in the healthy, questionable or even critical range. See also the picture below 😉

![image](https://github.com/UtilitechAS/amsreader-firmware/assets/6333880/8e20f750-4b2b-44f8-b72b-146117ce0f41)

**Reply by fkroepfl** [reply]
> We can se from your screenshots that meter serial number is read correctly, and so is Power and Energy.

Yes, that's how I see it so far ... hoping I haven't overlooked anything.

**Reply by fkroepfl** [reply]
By the way, do you have any additional information for me on how the date and time are coded?
I can decode or interpret all the other values for myself, but obviously I don't know enough about the coding of date and time. 😉

```
<!-- 28.05.2024 14:05:35 -->
  <OctetString Value="07E8051C020E0523FF800080" />
```

**Reply by ArnieO** [reply]
**Regarding reactive power**
As a private consumer you only pay for active power, which is why information on reactive power is of marginal (theoretical) interest for most of us. Large business consumers have some criteria on maximum cos(phi) (phase angle between active and reactive power). 
However, I fully agree with you that all information from the meter shall be displayed, which is why I asked you to check if it is visible for you.
You need to enable it on the config page:
![image](https://github.com/UtilitechAS/amsreader-firmware/assets/10295178/27047729-47a6-41d8-b32a-20f1d72e3531)
You will then see it displayed on the main page, like this:
![image](https://github.com/UtilitechAS/amsreader-firmware/assets/10295178/ba272fba-7f2f-4421-ae68-746a937d05f5)

**Coding of date and time**
```
<!-- 28.05.2024 14:05:35 -->
  <OctetString Value="07E8051C020E0523FF800080" />
```
I would expect this to be:
    07E8 (4 bytes): Represents the year
    051C (2 bytes): First byte is month, second is day
    020E (2 bytes): First byte hour, second byte minute
    05 (1 byte): Second
However, if we use this, we get     **28.05.2024 02:14:05** and not 28.05.2024 14:05:35 indicated above. 
So I'm not sure about the details here. Which version seems to be correct, I assume you know the approximate time of this payload?

**Reply by fkroepfl** [reply]
> You need to enable it on the config page:

thx, I'll check next time, when I'm on site 🤗

**Reply by fkroepfl** [reply]
> However, if we use this, we get **28.05.2024 02:14:05** and not 28.05.2024 14:05:35 indicated above.

🙏 

14 o'clock in the 24h view corresponds to the 2 o'clock display in the 12h view.
I cannot say where the deviation in minutes/seconds can come from.
We may be able to see more when I make or transmit the Telnet dump.

**Reply by fkroepfl** [reply]
> However, if we use this, we get **28.05.2024 02:14:05** and not 28.05.2024 14:05:35 indicated above.

If I follow your approach, but decode everything with 2 bytes after the year, then I find the time that the XML translator I used found. But then the question remains, what would the "02" stand for in this case? 

```
07e8 | ₍₁₆₎ | = | 2024 | ₍₁₀₎ year
  05 | ₍₁₆₎ | = |    5 | ₍₁₀₎ month
  1c | ₍₁₆₎ | = |   28 | ₍₁₀₎ day
  02 | ₍₁₆₎ | = |    2 | ₍₁₀₎ ?
  0e | ₍₁₆₎ | = |   14 | ₍₁₀₎ hh
  05 | ₍₁₆₎ | = |    5 | ₍₁₀₎ mm
  23 | ₍₁₆₎ | = |   35 | ₍₁₀₎ ss
  ff | ₍₁₆₎ | = |  255 | ₍₁₀₎ ?
  80 | ₍₁₆₎ | = |  128 | ₍₁₀₎ ?
  00 | ₍₁₆₎ | = |    0 | ₍₁₀₎ ?
  80 | ₍₁₆₎ | = |  128 | ₍₁₀₎ ?
```

🤷‍♂️ Or it can be assumed that it is an ASCII code 2:
- In the 7-bit ASCII character set, ASCII code 2 is represented by the control also known as the start of text (STX).

**Reply by ArnieO** [reply]
Great, I think you found it! 👍 
Just a guess, but maybe the "02" signifies time zone? Or indication for DST (summer vs winter time)?

**Reply by fkroepfl** [reply]
> Just a guess, but maybe the "02" signifies time zone? Or indication for DST (summer vs winter time)?

maybe, why not 👍 ... but not part of the time within the day 🤗

**Reply by fkroepfl** [reply]
Your approach that it has to do with the time difference makes perfect sense:

- Time zone in Vienna (GMT+2)

and x02 fits into the concept again. 😂

Is the entire hex string still COSEM/DLMS-compliant in this form? And if not, it is as it is. 🤷‍♂️

**Reply by ArnieO** [reply]
> Is the entire hex string still COSEM/DLMS-compliant in this form?

AFAIK it is not specified to this detail.

**Reply by gskjold** [reply]
The fourth field is day of week, 2 == tuesday. Timezone is at the end of the data (0x8000 == not specified), last field is status (0x80 == DST active enabled)
See blue book page 38 (https://github.com/UtilitechAS/amsreader-firmware/blob/main/doc/Blue_Book_Edition_13-Excerpt.pdf)

**Reply by fkroepfl** [reply]
Reading the documents, if you know which ones, can - as you can see - definitely help ... 😉🙏

**Reply by fkroepfl** [reply]
According to the referenced specification, the recorded data set then naturally reads like this and no other ... 😉 .
```
07e8 | ₍₁₆₎ | = |  2024 | ₍₁₀₎ year (highbyte|lowbyte)
  05 | ₍₁₆₎ | = |     5 | ₍₁₀₎ month 
  1c | ₍₁₆₎ | = |    28 | ₍₁₀₎ day of month
  02 | ₍₁₆₎ | = |     2 | ₍₁₀₎ day of week
  0e | ₍₁₆₎ | = |    14 | ₍₁₀₎ hour
  05 | ₍₁₆₎ | = |     5 | ₍₁₀₎ minute
  23 | ₍₁₆₎ | = |    35 | ₍₁₀₎ second
  ff | ₍₁₆₎ | = |   255 | ₍₁₀₎ hundredths of second (xFF=not specified)
8000 | ₍₁₆₎ | = | 32768 | ₍₁₀₎ deviation (highbyte|lowbyte; x8000=not specified)
  80 | ₍₁₆₎ | = |   128 | ₍₁₀₎ clock status (bit 7 MSB set = daylight saving active)
```

**Reply by fkroepfl** [reply]
> To help understand the reason for that, could you run a [Tenet debug](https://www.amsleser.no/module/ets_blog/blog?id_post=24), and post the result?

@ArnieO FYI, hope this is it, what you expect 

- [telnetdump.txt](https://github.com/user-attachments/files/15586949/telnetdump.txt)

**Reply by ArnieO** [reply]
That's it. Now let us see if @gskjold based on that can figure out why you get nothing on I(L1), I(L2 and I(L3). I can see they are included in the data, but the values seem to be a bit odd for all the parameters. Was it connected to the meter when this was captured?

This is the DLMS part translated to XML:
```
<DataNotification>
  <LongInvokeIdAndPriority Value="0008F7B4" />
  <!--2024-06-05 14:29:55-->
  <DateTime Value="07E80605FF0E1D37FF8000FF" />
  <NotificationBody>
    <DataValue>
      <Structure Qty="0F" >
        <Array Qty="0F" >
          <Structure Qty="04" >
            <UInt16 Value="0028" />
            <!--0.8.25.9.0.255-->
            <OctetString Value="0008190900FF" />
            <Int8 Value="02" />
            <UInt16 Value="0000" />
          </Structure>
          <Structure Qty="04" >
            <UInt16 Value="0028" />
            <!--0.8.25.9.0.255-->
            <OctetString Value="0008190900FF" />
            <Int8 Value="01" />
            <UInt16 Value="0000" />
          </Structure>
          <Structure Qty="04" >
            <UInt16 Value="0001" />
            <!--0.0.96.1.0.255-->
            <OctetString Value="0000600100FF" />
            <Int8 Value="02" />
            <UInt16 Value="0000" />
          </Structure>
          <Structure Qty="04" >
            <UInt16 Value="0008" />
            <!--0.0.1.0.0.255-->
            <OctetString Value="0000010000FF" />
            <Int8 Value="02" />
            <UInt16 Value="0000" />
          </Structure>
          <Structure Qty="04" >
            <UInt16 Value="0003" />
            <!--1.0.1.7.0.255-->
            <OctetString Value="0100010700FF" />
            <Int8 Value="02" />
            <UInt16 Value="0000" />
          </Structure>
          <Structure Qty="04" >
            <UInt16 Value="0003" />
            <!--1.0.2.7.0.255-->
            <OctetString Value="0100020700FF" />
            <Int8 Value="02" />
            <UInt16 Value="0000" />
          </Structure>
          <Structure Qty="04" >
            <UInt16 Value="0003" />
            <!--1.0.1.8.0.255-->
            <OctetString Value="0100010800FF" />
            <Int8 Value="02" />
            <UInt16 Value="0000" />
          </Structure>
          <Structure Qty="04" >
            <UInt16 Value="0003" />
            <!--1.0.2.8.0.255-->
            <OctetString Value="0100020800FF" />
            <Int8 Value="02" />
            <UInt16 Value="0000" />
          </Structure>
          <Structure Qty="04" >
            <UInt16 Value="0003" />
            <!--1.0.3.7.0.255-->
            <OctetString Value="0100030700FF" />
            <Int8 Value="02" />
            <UInt16 Value="0000" />
          </Structure>
          <Structure Qty="04" >
            <UInt16 Value="0003" />
            <!--1.0.4.7.0.255-->
            <OctetString Value="0100040700FF" />
            <Int8 Value="02" />
            <UInt16 Value="0000" />
          </Structure>
          <Structure Qty="04" >
            <UInt16 Value="0003" />
            <!--1.0.3.8.0.255-->
            <OctetString Value="0100030800FF" />
            <Int8 Value="02" />
            <UInt16 Value="0000" />
          </Structure>
          <Structure Qty="04" >
            <UInt16 Value="0003" />
            <!--1.0.4.8.0.255-->
            <OctetString Value="0100040800FF" />
            <Int8 Value="02" />
            <UInt16 Value="0000" />
          </Structure>
          <Structure Qty="04" >
            <UInt16 Value="0003" />
            <!--1.0.31.7.0.255-->
            <OctetString Value="01001F0700FF" />
            <Int8 Value="02" />
            <UInt16 Value="0000" />
          </Structure>
          <Structure Qty="04" >
            <UInt16 Value="0003" />
            <!--1.0.51.7.0.255-->
            <OctetString Value="0100330700FF" />
            <Int8 Value="02" />
            <UInt16 Value="0000" />
          </Structure>
          <Structure Qty="04" >
            <UInt16 Value="0003" />
            <!--1.0.71.7.0.255-->
            <OctetString Value="0100470700FF" />
            <Int8 Value="02" />
            <UInt16 Value="0000" />
          </Structure>
        </Array>
        <!--0.8.25.9.0.255-->
        <OctetString Value="0008190900FF" />
        <!--62854225-->
        <OctetString Value="3632383534323235" />
        <!--2024-06-05 14:29:55-->
        <OctetString Value="07E80605030E1D37FF800080" />
        <UInt32 Value="00000000" />
        <UInt32 Value="00000572" />
        <UInt32 Value="00BDCF90" />
        <UInt32 Value="001C7C0A" />
        <UInt32 Value="00000507" />
        <UInt32 Value="00000000" />
        <UInt32 Value="004DF056" />
        <UInt32 Value="000FFAC6" />
        <UInt16 Value="0153" />
        <UInt16 Value="0162" />
        <UInt16 Value="00E6" />
      </Structure>
    </DataValue>
  </NotificationBody>
</DataNotification>
```

**Reply by fkroepfl** [reply]
> I can see they are included in the data, but the values seem to be a bit odd for all the parameters. Was it connected to the meter when this was captured?

Yes, the module was connected to the meter. 

They don't look too bad in the XML translation, do they?

```
<UInt16 Value="0153" /> 0153 ₍₁₆₎ = 339 ₍₁₀₎ => 3,39 A ?
<UInt16 Value="0162" /> 0162 ₍₁₆₎ = 354 ₍₁₀₎ => 3,54 A ?
<UInt16 Value="00E6" /> 00e6 ₍₁₆₎ = 230 ₍₁₀₎ => 2,30 A ?

```

**Reply by fkroepfl** [reply]
Today's screenshot for your information ...

<img width="1483" alt="Bildschirmfoto 2024-06-05 um 14 42 58" src="https://github.com/UtilitechAS/amsreader-firmware/assets/6333880/0ec8f683-3b81-4905-a27c-4840cb35abd5">

**Reply by ArnieO** [reply]
> They don't look too bad in the XML translation, do they?

You're right - this looks fine. And thank you for screenshot!

**Reply by fkroepfl** [reply]
> By the way, do you have any additional information for me on how the date and time are coded? I can decode or interpret all the other values for myself, but obviously I don't know enough about the coding of date and time. 😉
> 
> ```
> 
>   <OctetString Value="07E8051C020E0523FF800080" />
> ```

Questions about the timestamp: Is it possible or is there possibly even a plan yet to display this value as well?

Background: I think it would be advantageous or useful for visual inspection if all the register values provided by a smart meter were visible in the app, wouldn't it?

**Comment by fkroepfl** [comment]
I have now also looked in the old logs and filtered out 10 values. It seems to confirm more and more 😉

```
YY YY MM DD    hh mm ss
07 e8 05 1c 02 0e 05 1e ff
07 e8 05 1c 02 0e 05 23 ff
07 e8 05 1c 02 0e 05 28 ff
07 e8 05 1c 02 0e 05 2d ff
07 e8 05 1c 02 0e 05 32 ff
07 e8 05 1c 02 0e 05 37 ff
07 e8 05 1c 02 0e 06 00 ff
07 e8 05 1c 02 0e 06 05 ff
07 e8 05 1c 02 0e 06 0a ff
07 e8 05 1c 02 0e 06 0f ff
```

**Comment by fkroepfl** [comment]
> As a private consumer you only pay for active power, which is why information on reactive power is of marginal (theoretical) interest for most of us. Large business consumers have some criteria on maximum cos(phi) (phase angle between active and reactive power).

You're right. However, the reactive power or reactive energy remains as a loss. And if you don't argue directly with money, but with the proportionate CO2 emissions and thus the sustainability of a household, then you could also make the statement:

- This is how "green" your household is per kilowatt hour consumed. 🤗

To quote Bill Gates "You're never going to get the amount of CO2 emitted to go down unless you deal with the one magic metric, which is CO2 per kilowatt-hour." 😉

**Comment by gskjold** [comment]
Please try the firmware below:
[esp32s3.zip](https://github.com/user-attachments/files/15633553/esp32s3.zip)
[esp32solo.zip](https://github.com/user-attachments/files/15633561/esp32solo.zip)
[esp32.zip](https://github.com/user-attachments/files/15633564/esp32.zip)
[esp8266.zip](https://github.com/user-attachments/files/15633567/esp8266.zip)
[esp32c3.zip](https://github.com/user-attachments/files/15633570/esp32c3.zip)
[esp32s2.zip](https://github.com/user-attachments/files/15633572/esp32s2.zip)

Flashing instructions: https://github.com/UtilitechAS/amsreader-firmware/wiki/Flashing-via-web-GUI

**Reply by fkroepfl** [reply]
I'll be happy to do it when I'm back on site.
What should I look out for? What has changed?

**Reply by ArnieO** [reply]
Your issue was the missing amperage values - right?

**Reply by fkroepfl** [reply]
OK, all right.

**Reply by fkroepfl** [reply]
FYI: No change visible in the amperage display in the dashboard. I uploaded the firmware from the ESP32.zip file.
Question: Should the new firmware have a different or the same release number?

<img width="1233" alt="Bildschirmfoto 2024-06-07 um 13 37 27" src="https://github.com/UtilitechAS/amsreader-firmware/assets/6333880/f37ae3b5-f00a-498d-a98c-97db62f6a8ac">
<img width="1233" alt="Bildschirmfoto 2024-06-07 um 13 40 47" src="https://github.com/UtilitechAS/amsreader-firmware/assets/6333880/c07f60e6-53d4-452d-8ffe-abc281b7e704">
<img width="1233" alt="Bildschirmfoto 2024-06-07 um 13 43 21" src="https://github.com/UtilitechAS/amsreader-firmware/assets/6333880/a06f3676-efb4-498e-98e6-d81ffc6c2539">

**Reply by ArnieO** [reply]
No, it should look like this:
![image](https://github.com/UtilitechAS/amsreader-firmware/assets/10295178/40f8d5f8-f48c-46ed-b452-d4ec0f7a4a05)

Please retry.

**Reply by fkroepfl** [reply]
sorry, but even after factory reset, new download of your posted FW, it ends up with the same FW release

<img width="250" alt="Bildschirmfoto 2024-06-07 um 14 57 39" src="https://github.com/UtilitechAS/amsreader-firmware/assets/6333880/c6aab841-76f0-412f-a93f-95cd777b328c">

**Reply by fkroepfl** [reply]
Now it is done, but only after a complete "ERASE" and a new flashing with the https://adafruit.github.io/Adafruit_WebSerial_ESPTool/

`Installed version: 5eeab2b
`

I have tried https://github.com/UtilitechAS/amsreader-firmware/wiki/Flashing-via-web-GUI twice, but without success.

As I'm no longer on site, I can only test the new FW on my next visit. Sorry about that.

**Reply by ArnieO** [reply]
OK, so you managed to update it; good!

**Reply by fkroepfl** [reply]
Looks good.

Question about the current values: how are they or should they be displayed during export or is the current direction not taken into account?

<img width="643" alt="Bildschirmfoto 2024-06-12 um 15 54 53" src="https://github.com/UtilitechAS/amsreader-firmware/assets/6333880/f3b11668-0719-46a8-9c16-3072160aae0f">

**Reply by fkroepfl** [reply]
It may be possible to optimize the labeling of the data columns somewhat 😉.

<img width="1300" alt="Bildschirmfoto 2024-06-12 um 16 11 46" src="https://github.com/UtilitechAS/amsreader-firmware/assets/6333880/c53334d2-ef22-4b7e-a0c3-8c18f72d5cfb">

**Reply by ArnieO** [reply]
Maybe, but please be more specific! 
Also, you may be interested in changing UI language to German?

Step 1: Load from server:
![image](https://github.com/UtilitechAS/amsreader-firmware/assets/10295178/7ad8e518-3c00-49ab-b1f1-cce623032a92)

Step 2: Select language:
![image](https://github.com/UtilitechAS/amsreader-firmware/assets/10295178/63e3ff7a-0288-42d2-8843-1c3a63f41999)

**Reply by ArnieO** [reply]
> Question about the current values: how are they or should they be displayed during export or is the current direction not taken into account?

You configure for production by setting the peak power of your installation here:
![image](https://github.com/UtilitechAS/amsreader-firmware/assets/10295178/c91c503b-8aa4-48fa-80e6-4340acd54127)

The current display will then change, showing negative values for export. This user has export configured, but was importing when screenclip was taken:
![image](https://github.com/UtilitechAS/amsreader-firmware/assets/10295178/3b3b23ce-1f36-4006-9ba3-1e11c20364ca)

**Reply by fkroepfl** [reply]
> Maybe, but please be more specific!

I meant, for example, that the label could always be placed at a certain distance above the data column and not mixed, once above the data column and once in the data column. 

This approach seems to be more safe in that the probability that one or all phases are 100% utilized is rather low, but that there are low current values is very likely. 🤗

![image](https://github.com/UtilitechAS/amsreader-firmware/assets/6333880/705db596-9ca4-48a6-bd23-d4f68e953856)

**Reply by fkroepfl** [reply]
> Also, you may be interested in changing UI language to German?

Definitely, I'll do that when I'm done with all the questions and points I still have to work on 😉 .

**Reply by ArnieO** [reply]
If the bars are low, the numbers are like this:
![image](https://github.com/UtilitechAS/amsreader-firmware/assets/10295178/3288eea3-7b73-4dfd-b5b1-97454c334bb9)

**Reply by fkroepfl** [reply]
> If the bars are low, the numbers are like this:

OK, I see, but then a small inconsistency may have crept into the code. 😉

<img width="269" alt="Bildschirmfoto 2024-06-13 um 16 21 34" src="https://github.com/UtilitechAS/amsreader-firmware/assets/6333880/f8bde409-1f06-47aa-9ed2-7ad7d2ddf341">
